# Credit Card Fraud Detection - Data Analysis Layer

*By Satyam Sharma, B.Tech CSE(ML&AI), 5th Sem, Section ML, Graphic Era Deemed University*


### Context

It is important that credit card companies are able to recognize fraudulent credit card transactions so that customers are not charged for items that they did not purchase.

## Content of the Data Set

The datasets contains transactions made by credit cards in September 2013 by european cardholders.
This dataset presents transactions that occurred in two days, where we have 492 frauds out of 284,807 transactions. The dataset is highly unbalanced, the positive class (frauds) account for 0.172% of all transactions.

It contains only numerical input variables which are the result of a PCA transformation. Unfortunately, due to confidentiality issues, we cannot provide the original features and more background information about the data. Features V1, V2, … V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-senstive learning. Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.

Let us explore the dataset in a bit more detail.

---

First, lets import the libraries required to work on the data.

In [18]:
import numpy as np
import pandas as pd

Then import the dataset itself into an object

In [3]:
df = pd.read_csv('creditcard.csv')

Now lets start looking into the data - reading the dimensions of the dataset using its shape var, then getting a general idea about its composition using head() & describe().

In [6]:
print(df.shape)

(284807, 31)


In [8]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [9]:
df.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  3.919560e-15  5.688174e-16 -8.769071e-15  2.782312e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean  -1.552563e-15  2.010663e-15 -1.694249e-15 -1.927028e-16 -3.137024e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.537294e-16  7.959909e-16  5.367590e-16  4.458112e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   1.453003e-15  1.699104e-15 -3.660161e-16 -1.206049e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

We are dealing with problems which lead to a resultant of either true(fraudulent transaction) or false(legit transaction). In the given dataset, these results are stored in the 'Class' row, but when we pay attention to the describe() resultant, we can see that the composition of the results in 'Class' is quite skewed, given its mean of just 0.0017~. 

Let us take a closer look at this. 

In [19]:
fraud = df[df['Class'] == 1] 
legit = df[df['Class'] == 0] 

print('Number of fraudulent cases = {}'.format(len(df[df['Class'] == 1]))) 
print('Number of legit cases = {}'.format(len(df[df['Class'] == 0])))

outlierRatio = len(fraud)/float(len(legit)) 
print("Percentage of fraudulent cases = ",outlierRatio*100,"%") 

Number of fraudulent cases = 492
Number of legit cases = 284315
Percentage of fraudulent cases =  0.17304750013189596 %


---
Given this result, ie just a 0.17304750013189596 % chance of the transaction begin fraudulent(1), we can hence say that we are dealing with an imbalance in the data.

Also, from the details of the Amount row of describe(), we can also see that the average money transaction for the fraudulent ones is more. This makes this problem even more crucial to deal with.

## Result of the Data Analysis

An imbalance in data has been found, and hence we can classify this problem as an imbalanced classification problem. 

There are several ways to deal with this problem.

**Change the performance metric** - Using other means to calcuate/understand the accuracy of the model. Eg :-

Confusion Matrix: showing correct predictions and types of incorrect predictions.

Precision: The number of true positives divided by all positive predictions. Low precision indicates a high number of false positives.

Recall: The number of true positives divided by the number of positive values in the test data. Low recall indicates a high number of false negatives.

F1 Score: the weighted average of precision and recall.

AUC: is a graph showing the performance of a classification model at all classification thresholds.

**Random Oversampling** - By 'oversampling', ie balancing the data set by copying minority class samples. Like Random Oversampling. Its the addition of randomly selected samples from the minority class. This technique can be used if the data set is small. It may cause overfitting.

**SMOTE Oversampling** - Creating synthetic samples from minority class to prevent overfitting.

SMOTE first selects a minority class instance a at random and finds its k nearest minority class neighbors. The synthetic instance is then created by choosing one of the k nearest neighbors b at random and connecting a and b to form a line segment in the feature space. The synthetic instances are generated as a convex combination of the two chosen instances a and b.

**Random Undersampling** - Undersampling is the technique of balancing the data set by removing the samples belonging to the majority class. You can use this technique if you have a large data set. In Random Undersampling, information may be lost due to random selection.

**NearMiss Undersampling** - It prevents the loss of information. It is based on the KNN algorithm. The distance between the samples belonging to the majority class and the samples belonging to the minority class is calculated. Samples whose distance is shorter than the specified value of k are preserved.

Three Version:

NearMiss-1: Majority class examples with minimum average distance to three closest minority class examples.
NearMiss-2: Majority class examples with minimum average distance to three furthest minority class examples.
NearMiss-3: Majority class examples with minimum distance to each minority class example.
The NearMiss-3 seems desirable, given that it will only keep those majority class examples that are on the decision boundary

---

It is suggested that the Random Forest classifer be used to get a better results, as from Random Forest we can get which features are more important.